### ISSUE TITLE
Publish lychee for common platforms

### ISSUE THREAD

**Issue by mre** [lycheeverse/lychee#59]
## Introduction

lychee is getting more mature and we'd like to share the future v1 with a broader audience.
For that we need an easy way to install the binary.

Required platform support (roughly ordered by complexity):

* [X] cargo install
* [ ] Docker Hub (see #40 and #52)
* [ ] Binary releases on Github
* [ ] Homebrew for macOS
* [ ] Chocolatey for Windows
* [ ] apt (Debian/Ubuntu)

(This list is probably incomplete and should be extended.)

## How to contribute

Please pick one of the above targets and create a pull request.

Inspiration:
* Binary releases on Github: https://github.com/BurntSushi/ripgrep/blob/master/.github/workflows/release.yml
* APT package: https://github.com/sharkdp/bat/blob/master/.github/workflows/CICD.yml
* Chocolatey: https://github.com/marketplace/actions/chocolatey-action

**Comment by pawroman** [comment]
I'm happy to contribute an Arch Linux AUR package :+1:

**Comment by pawroman** [comment]
General remark: arch-wise, we could consider publishing binaries for x86_64, arm and aarch64 (64-bit ARM).

I'm sure there would be at least some ARM users.

**Comment by mre** [comment]
Sure! That would be great. Updated the list accordingly.

**Comment by pawroman** [comment]
Sorry for confusion, by this

> General remark: arch-wise, we could consider publishing binaries for x86_64, arm and aarch64 (64-bit ARM).

I meant that all of our packages should have versions for these CPU architectures. This includes Docker, binaries published on GitHub and other binary packages.

Arch Linux is just Arch Linux :)

**Comment by mre** [comment]
Noted. Updated the list accordingly. 👍

**Comment by mre** [comment]
Talked to @pizzamig and he'd be willing to support us with the FreeBSD package. Thanks!

**Comment by mre** [comment]
Hum... working on binary releases but it seems to be harder than expected. 
Developing this on a fork as I never get those CI builds right on first try. 😆 

Current pipeline:  
https://github.com/mre-early-access/lychee/blob/master/.github/workflows/release.yml

Last build:
https://github.com/mre-early-access/lychee/runs/1500158899?check_suite_focus=true

If someone wants to hack on this, go ahead.

**Comment by mre** [comment]
Binary releases are done. 🥳

**Comment by mre** [comment]
lychee Github action: https://github.com/marketplace/actions/lychee-link-checker-action

**Comment by alebcay** [comment]
Hi there 👋, I've submitted a pull request for including `lychee` in the Homebrew core repository at https://github.com/Homebrew/homebrew-core/pull/70216.

Just wanted to verify one thing - is `lychee` licensed under _either of_ (at the user's choice) `Apache-2.0` or `MIT`? This was my understanding although I could not find it explicitly stated anywhere.

**Comment by mre** [comment]
Oh that's so nice of you @alebcay. 🥳 Let's hope it gets accepted. 🤞 

> Just wanted to verify one thing - is lychee licensed under either of (at the user's choice) Apache-2.0 or MIT? This was my understanding although I could not find it explicitly stated anywhere.

Yes, exactly. I've added a paragraph to the end of the `README.md` now to clarify that.

**Comment by mre** [comment]
Looks like the build failed because `ring` couldn't build:

> error: unknown target triple 'unknown-apple-macosx11.0.0', please use -triple or -arch

That seems to be an issue with macOS Big Sur... 🤔 I wonder if that's supported by ring already.

**Comment by mre** [comment]
Some hints here: https://github.com/ogham/exa/issues/780

**Comment by mre** [comment]
https://github.com/Homebrew/homebrew-core/issues/68301

**Comment by orhun** [comment]
lychee is available in Arch Linux community repository: https://archlinux.org/packages/community/x86_64/lychee-link-checker/

**Comment by pawroman** [comment]
Issues/PRs blocking homebrew:
- https://github.com/softprops/hubcaps/issues/288
- https://github.com/async-email/async-smtp/pull/36 (needed by `check-if-email-exists`)

**Comment by fauust** [comment]
Hi! I am happy to help on the Debian/Ubuntu part. But I am not sure about what you want.

Do you want to produce deb packages in CI and make them available as assets (isn't the bin tar sufficient?) or are you interested in having lychee available directly on Debian/Ubuntu repositories?

**Comment by mre** [comment]
We'd love to have lychee available on Debian/Ubuntu repositories directly. 

```
apt get install lychee
```

would be so lit.

**Comment by mre** [comment]
And thanks for offering your help. Any support here is highly appreciated.

**Comment by fauust** [comment]
:)
Ok, but before `apt-get install lychee` is possible, lot (lot) needs to be done, Debian packaging represent lots of work...

Anyway, as in your [bat example](https://github.com/sharkdp/bat/blob/master/.github/workflows/CICD.yml), does it make sense to propose .deb files as assets? Also, for debian/ubuntu packaging, the first thing that comes to my mind and that is missing is a manual page.

**Comment by mre** [comment]
Okay a manpage should be doable. Think clap has a generator of some sorts for that -- at least an unofficial one.
Creating self-hosted debian packages would be an option for now, too. 👍

**Comment by mre** [comment]
@alebcay, thanks so much for adding lychee to homebrew.

```
brew install lychee
```

works as expected. ✌️

**Comment by thunder-coding** [comment]
I have added `lychee` for [Termux](https://termux.org), see termux/termux-packages#7050 🚀🚀🎉

**Comment by mre** [comment]
Great! Can you create a PR to mention that in the README? ✨

**Comment by thunder-coding** [comment]
> Great! Can you create a PR to mention that in the README? ✨

See #274

**Comment by mre** [comment]
Thanks!

**Comment by fauust** [comment]
> Okay a manpage should be doable. Think clap has a generator of some sorts for that -- at least an unofficial one.
> Creating self-hosted debian packages would be an option for now, too. +1

Hey @mre!
I may have some spare time in the next weeks to look at Debian packaging.

So, I have some questions to understand better in which direction you want to go:
- what about creating a  `debian/` directory with all the necessary stuff, wouldn't that be easier to maintain than having everything generated in the CI (like in the bat example you are pointing to)?
- do you have any reason why you prefer cargo-deb over [debcargo](https://salsa.debian.org/rust-team/debcargo) (I do not have experience with those tools BTW)?

Cheers!

**Comment by mre** [comment]
Thanks for helping out. debcargo looks nice. Let's try it!
I'm fine with adding a debian subfolder. Reading through the debcargo docs, we might only need one toml file anyway, which would be sweet.

**Comment by fauust** [comment]
@mre I am still not sure about the best solution. bat is packaged into Debian so I may contact the maintainer to understand what workflow he use and then decide based on that.

My idea is to reduce as much as possible the diff (or the sync mechanism) between upstream and downstream Debian/Ubuntu packaging.

**Comment by mre** [comment]
> bat is packaged into Debian so I may contact the maintainer to understand what workflow he use and then decide based on that.

Sounds like a good idea. Thanks for looking into this.
Also perhaps you can give debcargo a spin and see what output it produces to have a starting point.

**Comment by fauust** [comment]
Working on another packaging project (Go), I found out where bat is packaged for Debian:
https://salsa.debian.org/rust-team/debcargo-conf/-/tree/master/src/bat/debian

GIven the number of Rust packages that seems to be packaged this way, I believe this is the way to go:
https://salsa.debian.org/rust-team/debcargo-conf/-/tree/master/src/

So, next step is to understand better the workflow...

**Comment by san-ouadghiri** [comment]
Hi everyone,
Have-we consider making lychee available on the npm registry? Could we ahd this one to the list?
`npm install lychee --save-dev`

**Comment by mre** [comment]
Yes, we discussed that in https://github.com/lycheeverse/lychee/issues/420.
I'll add it to the list. I think we can use [Neon](https://neon-bindings.com/) to generate the bindings. This could start as a separate project in a separate repo, so if anyone wants to give it a try, feel free. I want to tackle this after 1.0, probably, but maybe someone finds the time to experiment with it and report back.

**Comment by mre** [comment]
We need someone that can package lychee for Windows with Chocolatey and winget? Any takers? 😊

**Comment by lebensterben** [comment]
@mre 
We may also add lychee to https://github.com/alsuren/cargo-quickinstall

**Comment by mre** [comment]
Yeah that's a great idea. Let me add that to the list

**Comment by soredake** [comment]
https://github.com/chocolatey-community/chocolatey-package-requests/issues/1391
https://github.com/microsoft/winget-pkgs/issues/91395

**Comment by mre** [comment]
Amazing @soredake. Thanks so much for pushing this forward.

**Comment by mre** [comment]
Worked like a charm.

Needed to update to the latest winget version. With the old version (1.2) I got a confusing error message at first:

```
> winget install lychee
winget 0x80070032 : The request is not supported.
```

Updated winget through the Microsoft Store by updating "App Installer", which contains winget.
Then from an **administrator** power shell:

```
PS C:\Windows\system32> winget -v
v1.3.2691

PS C:\Windows\system32> winget install lychee
Found lychee [lycheeverse.lychee] Version 0.10.3
This application is licensed to you by its owner.
Microsoft is not responsible for, nor does it grant any licenses to, third-party packages.
Downloading https://github.com/lycheeverse/lychee/releases/download/v0.10.3/lychee-v0.10.3-windows-x86_64.exe
  ██████████████████████████████  12.7 MB / 12.7 MB
Successfully verified installer hash
Starting package install...
Overwriting existing file: C:\Users\mendler\AppData\Local\Microsoft\WinGet\Links\lychee.exe
Successfully installed
```

Closed and reopened the shell and lychee was in my `PATH` and worked as expected:

```
PS C:\Windows\system32> Get-Command lychee

CommandType     Name                                               Version    Source
-----------     ----                                               -------    ------
Application     lychee.exe                                         0.0.0.0    C:\Users\mendler\AppData\Local\Microsoft\WinGet\Links\lychee.exe

PS C:\Windows\system32>  lychee ...
```

Thanks a lot @soredake.

**Comment by lebensterben** [comment]
asdf is not a package system designed for distributing packages. it's only designed as a version manager for programming language toolchains.

the entire asdf-plugin ecosystem is abusing it.

**Comment by aj-stein-nist** [comment]
> asdf is not a package system designed for distributing packages. it's only designed as a version manager for programming language toolchains.
> 
> the entire asdf-plugin ecosystem is abusing it.

So I guess I can take that as a no vote. No worries, I just thought it would be usefully for testing parallel versions simultaneously without installing (some plugins existing to manage binaries that are not strictly for programming language runtimes). It would have been useful for testing issue 1020, but no worries.

**Comment by mre** [comment]
Yeah, I have similar feelings about asdf as @lebensterben. Always saw it as a version manager for programming languages &mdash; not programs. In any case, thanks for the suggestion @aj-stein-nist.

**Comment by mre** [comment]
Of course. Go ahead. 👍 Happy hacking.

**Comment by echoix** [comment]
I also see [cargo-binstall](https://github.com/cargo-bins/cargo-binstall) as an easy target, since multiple binaries for multiple target triplets are already in the GitHub releases (it is really useful to have so many, especially for musl and arm, where compiling on arm devices can be long!).
Adding support for it is a matter of adding metadata in the Cargo.toml file, like described in :
https://github.com/cargo-bins/cargo-binstall/blob/main/SUPPORT.md

Cargo binstall helps to install binaries by simply downloading them from the releases, following the correct file naming for their platform, with a fallback on quickinstall, and a fallback of simply installing with cargo if nothing better was available.

**Comment by orhun** [comment]
Added `cargo-binstall` metadata in #1073

**Comment by mre** [comment]
Awesome job! `cargo-binstall` support got merged now. 🥳

**Comment by soredake** [comment]
Can [snapcraft](https://snapcraft.io/) / [flatpak](https://flatpak.org/) be added to list?

**Comment by mre** [comment]
Sure, done

**Comment by paolobarbolini** [comment]
I've fixed the tag format (newish tags changing from `v[version]` to `lychee-v[version]`) at https://release-monitoring.org so that distros/maintainers who rely on it can find out about new releases and get the new versions out quickly.

I've found Alpine Linux, NixOS (maintainers don't seem a lot active), FreshPorts (ping @pizzamig) have outdated versions. Alpine should pick it up with the next sync with Anitya, I'm not sure how it works for the other two.

**Comment by mre** [comment]
Thanks, Paolo.

**Comment by puneet-ekline** [comment]
Is it expected that I am unable to pin to a specific version of lychee using homebrew on MacOS

```
 ~ brew install lychee@0.18.0
==> Downloading https://formulae.brew.sh/api/formula.jws.json
################################################################################################################################################################ 100.0%
==> Downloading https://formulae.brew.sh/api/cask.jws.json
################################################################################################################################################################ 100.0%
Warning: No available formula with the name "lychee@0.18.0".
==> Searching for similarly named formulae and casks...
Error: No formulae or casks found for lychee@0.18.0.
```

**Comment by mre** [comment]
Honestly, I don't know how homebrew's version pinning mechanism works.

I was of the opinion that it would happen automatically based on the git tags. The homebrew file is also open source if you would like to take a look.

If some can provide more information, I'd be thankful for a comment.

**Comment by eread** [comment]
> Yeah, I have similar feelings about asdf as [@lebensterben](https://github.com/lebensterben). Always saw it as a version manager for programming languages — not programs. In any case, thanks for the suggestion [@aj-stein-nist](https://github.com/aj-stein-nist).

@mre @lebensterben @aj-stein-nist I wonder what you think of https://mise.jdx.dev?

It supports `asdf` plugins, but also more: https://mise.jdx.dev/registry.html. It is quite a cross-platform option for managing pinned-versions of installations. CC @puneet-ekline.

**Comment by mre** [comment]
Never heard of it, but it looks okay on first glance. Feel free to submit lychee there if you use mise. Maybe someone else already has experiences with it.

**Comment by eread** [comment]
> Never heard of it, but it looks okay on first glance. Feel free to submit lychee there if you use mise. Maybe someone else already has experiences with it.

@mre Raised: https://github.com/jdx/mise/pull/4181

**Comment by mre** [comment]
Oh nice. Either way is fine with me, but as you said, it's easier for users if it were on the `main` tree, so I'd prefer that. Thanks for looking into it.

**Comment by eread** [comment]
> > Never heard of it, but it looks okay on first glance. Feel free to submit lychee there if you use mise. Maybe someone else already has experiences with it.
> 
> [@mre](https://github.com/mre) Raised: [jdx/mise#4181](https://github.com/jdx/mise/pull/4181)

@mre with this release: https://github.com/jdx/mise/releases/tag/v2025.2.7, users can install and manage Lychee by using `mise` 👍🏻

**Comment by orhun** [comment]
After wrestling with TLS a bit, the Alpine package has been finally updated.

**Comment by mre** [comment]
The man, the legend. Thanks mate.

**Comment by egpbos** [comment]
For completeness sake: there is also a conda package on conda-forge: https://github.com/conda-forge/lychee-feedstock

**Comment by mre** [comment]
Ah, very cool. Can you create a pull request to mention it in the `README.md`?

**Comment by mre** [comment]
@steinbro added snap support! Thanks! 🥳 

```
$ snap install --edge lychee
lychee (edge) 0.19.1.dev from Matthias Endler (m-r-e) installed
$ lychee lychee/README.md 
   [WARN ] Error creating request: InvalidPathToUri("/CONTRIBUTING.md")
  111/111 ━━━━━━━━━━━━━━━━━━━━ Finished extracting links                        
🔍 111 Total (in 1s) ✅ 111 OK 🚫 0 Errors
```

**Comment by mre** [comment]
Guess we have to be a bit patient about the stable release. It won't let me choose the existing tags when triggering the workflow:

<img width="328" height="276" alt="Image" src="https://github.com/user-attachments/assets/6e9826cc-db18-4a15-9964-fece4d5c5f37" />

I'm assuming that's because the workflow didn't exist when the tag was created.

**Comment by egpbos** [comment]
> Ah, very cool. Can you create a pull request to mention it in the `README.md`?

Done :) #1822

**Comment by mre** [comment]
So cool! Thanks. 😊 Can you also add it here? https://github.com/lycheeverse/lycheeverse.github.io/edit/master/src/content/docs/installation.mdx 😉

**Comment by van-sprundel** [comment]
Would the team be open to a `cargo binstall` ? I'm not too familiar with `cargo quickinstall`, but it seems binstall is kind of the de-facto

**Comment by mre** [comment]
Yeah, that would be fantastic.
